# Hackathon: Predicting Custoemr Satisfaction from train ride

## Import Libraries

In [144]:
import numpy as np
import pandas as pd
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import LabelEncoder, OrdinalEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import StandardScaler


## Load Data

In [2]:
# Import survey test and train data
survey_train_data = pd.read_csv('Surveydata_train_.csv')
survey_test_data = pd.read_csv('Surveydata_test_.csv')

# Import travel test and train data
travel_train_data = pd.read_csv('Traveldata_train_.csv')
travel_test_data = pd.read_csv('Traveldata_test_.csv')

# Data Overview

In [3]:
survey_train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 94379 entries, 0 to 94378
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   ID                       94379 non-null  int64 
 1   Overall_Experience       94379 non-null  int64 
 2   Seat_Comfort             94318 non-null  object
 3   Seat_Class               94379 non-null  object
 4   Arrival_Time_Convenient  85449 non-null  object
 5   Catering                 85638 non-null  object
 6   Platform_Location        94349 non-null  object
 7   Onboard_Wifi_Service     94349 non-null  object
 8   Onboard_Entertainment    94361 non-null  object
 9   Online_Support           94288 non-null  object
 10  Ease_of_Online_Booking   94306 non-null  object
 11  Onboard_Service          86778 non-null  object
 12  Legroom                  94289 non-null  object
 13  Baggage_Handling         94237 non-null  object
 14  CheckIn_Service          94302 non-nul

**Observations:** There are 17 columns in total with 94,379 rows. Some of the columns are missing values. There are two types of columns (int(2) and object(15)). 

In [4]:
survey_test_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35602 entries, 0 to 35601
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   ID                       35602 non-null  int64 
 1   Seat_Comfort             35580 non-null  object
 2   Seat_Class               35602 non-null  object
 3   Arrival_Time_Convenient  32277 non-null  object
 4   Catering                 32245 non-null  object
 5   Platform_Location        35590 non-null  object
 6   Onboard_Wifi_Service     35590 non-null  object
 7   Onboard_Entertainment    35594 non-null  object
 8   Online_Support           35576 non-null  object
 9   Ease_of_Online_Booking   35584 non-null  object
 10  Onboard_Service          32730 non-null  object
 11  Legroom                  35577 non-null  object
 12  Baggage_Handling         35562 non-null  object
 13  CheckIn_Service          35580 non-null  object
 14  Cleanliness              35600 non-nul

**Observations:** There are 16 columns (Overall_Experience is missing from training data) and 35,602 rows. There are missing values in training data. There are two types of columns int(1) and object(15).

In [5]:
travel_train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 94379 entries, 0 to 94378
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       94379 non-null  int64  
 1   Gender                   94302 non-null  object 
 2   Customer_Type            85428 non-null  object 
 3   Age                      94346 non-null  float64
 4   Type_Travel              85153 non-null  object 
 5   Travel_Class             94379 non-null  object 
 6   Travel_Distance          94379 non-null  int64  
 7   Departure_Delay_in_Mins  94322 non-null  float64
 8   Arrival_Delay_in_Mins    94022 non-null  float64
dtypes: float64(3), int64(2), object(4)
memory usage: 6.5+ MB


**Observations:** There are 9 columns with 94,379 rows. There are three types of columns int(2), float(3), and object(4). There are missing values in some of the columns.

In [6]:
travel_test_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35602 entries, 0 to 35601
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       35602 non-null  int64  
 1   Gender                   35572 non-null  object 
 2   Customer_Type            32219 non-null  object 
 3   Age                      35591 non-null  float64
 4   Type_Travel              32154 non-null  object 
 5   Travel_Class             35602 non-null  object 
 6   Travel_Distance          35602 non-null  int64  
 7   Departure_Delay_in_Mins  35573 non-null  float64
 8   Arrival_Delay_in_Mins    35479 non-null  float64
dtypes: float64(3), int64(2), object(4)
memory usage: 2.4+ MB


**Observations:** There are 9 columns with 35,602 rows. There are three types of columns int(2), float(3), and object(4). There are missing values in some of the columns.

In [8]:
survey_train_data.describe().T

count          mean           std         min  \
ID                  94379.0  9.884719e+07  27245.014865  98800001.0   
Overall_Experience  94379.0  5.466576e-01      0.497821         0.0   

                           25%         50%         75%         max  
ID                  98823595.5  98847190.0  98870784.5  98894379.0  
Overall_Experience         0.0         1.0         1.0         1.0

In [9]:
survey_test_data.describe().T

count        mean           std         min          25%         50%  \
ID  35602.0  99917801.5  10277.556478  99900001.0  99908901.25  99917801.5   

            75%         max  
ID  99926701.75  99935602.0

**Observations:** There is nothing gotten from describe on the survey data since most variables are objects

In [10]:
travel_train_data.describe().T

count          mean           std         min  \
ID                       94379.0  9.884719e+07  27245.014865  98800001.0   
Age                      94346.0  3.941965e+01     15.116632         7.0   
Travel_Distance          94379.0  1.978888e+03   1027.961019        50.0   
Departure_Delay_in_Mins  94322.0  1.464709e+01     38.138781         0.0   
Arrival_Delay_in_Mins    94022.0  1.500522e+01     38.439409         0.0   

                                25%         50%         75%         max  
ID                       98823595.5  98847190.0  98870784.5  98894379.0  
Age                            27.0        40.0        51.0        85.0  
Travel_Distance              1359.0      1923.0      2538.0      6951.0  
Departure_Delay_in_Mins         0.0         0.0        12.0      1592.0  
Arrival_Delay_in_Mins           0.0         0.0        13.0      1584.0

In [11]:
travel_test_data.describe().T

count          mean           std         min  \
ID                       35602.0  9.991780e+07  10277.556478  99900001.0   
Age                      35591.0  3.944700e+01     15.137554         7.0   
Travel_Distance          35602.0  1.987152e+03   1024.308863        50.0   
Departure_Delay_in_Mins  35573.0  1.488070e+01     37.895453         0.0   
Arrival_Delay_in_Mins    35479.0  1.530880e+01     38.531293         0.0   

                                 25%         50%          75%         max  
ID                       99908901.25  99917801.5  99926701.75  99935602.0  
Age                            27.00        40.0        51.00        85.0  
Travel_Distance              1360.00      1929.0      2559.00      6868.0  
Departure_Delay_in_Mins         0.00         0.0        13.00       978.0  
Arrival_Delay_in_Mins           0.00         0.0        13.00       970.0

**Observations:** We can see that age might have a little skewness, Travel_Distance will have outliers as our max values is much larger than even Q3, Departure_Delay_in_Mins will have outliers as the max value is much larger so it will be skewed, Arrival_Delay_in_Mins will be skewed right as the max value is much larger than most other values.

In [40]:
survey_train_data.isna().sum()

ID                            0
Overall_Experience            0
Seat_Comfort                 61
Seat_Class                    0
Arrival_Time_Convenient    8930
Catering                   8741
Platform_Location            30
Onboard_Wifi_Service         30
Onboard_Entertainment        18
Online_Support               91
Ease_of_Online_Booking       73
Onboard_Service            7601
Legroom                      90
Baggage_Handling            142
CheckIn_Service              77
Cleanliness                   6
Online_Boarding               6
dtype: int64

In [42]:
survey_test_data.isna().sum()

ID                            0
Seat_Comfort                 22
Seat_Class                    0
Arrival_Time_Convenient    3325
Catering                   3357
Platform_Location            12
Onboard_Wifi_Service         12
Onboard_Entertainment         8
Online_Support               26
Ease_of_Online_Booking       18
Onboard_Service            2872
Legroom                      25
Baggage_Handling             40
CheckIn_Service              22
Cleanliness                   2
Online_Boarding               2
dtype: int64

In [43]:
travel_train_data.isna().sum()

ID                            0
Gender                       77
Customer_Type              8951
Age                          33
Type_Travel                9226
Travel_Class                  0
Travel_Distance               0
Departure_Delay_in_Mins      57
Arrival_Delay_in_Mins       357
dtype: int64

In [44]:
travel_test_data.isna().sum()

ID                            0
Gender                       30
Customer_Type              3383
Age                          11
Type_Travel                3448
Travel_Class                  0
Travel_Distance               0
Departure_Delay_in_Mins      29
Arrival_Delay_in_Mins       123
dtype: int64

In [23]:
def get_value_counts(df):
    for col in df.columns:
        print(f"Value counts for column: {col}")
        print("-----------------------------------------")
        print(df[col].value_counts())
        print("\n\n")

In [24]:
get_value_counts(survey_train_data)

Value counts for column: ID
-----------------------------------------
98800001    1
98862916    1
98862925    1
98862924    1
98862923    1
           ..
98831459    1
98831458    1
98831457    1
98831456    1
98894379    1
Name: ID, Length: 94379, dtype: int64



Value counts for column: Overall_Experience
-----------------------------------------
1    51593
0    42786
Name: Overall_Experience, dtype: int64



Value counts for column: Seat_Comfort
-----------------------------------------
Acceptable           21158
Needs Improvement    20946
Good                 20595
Poor                 15185
Excellent            12971
Extremely Poor        3463
Name: Seat_Comfort, dtype: int64



Value counts for column: Seat_Class
-----------------------------------------
Green Car    47435
Ordinary     46944
Name: Seat_Class, dtype: int64



Value counts for column: Arrival_Time_Convenient
-----------------------------------------
Good                 19574
Excellent            17684
Acceptable  

In [25]:
get_value_counts(survey_test_data)

Value counts for column: ID
-----------------------------------------
99900001    1
99923738    1
99923732    1
99923733    1
99923734    1
           ..
99911862    1
99911861    1
99911860    1
99911859    1
99935602    1
Name: ID, Length: 35602, dtype: int64



Value counts for column: Seat_Comfort
-----------------------------------------
Acceptable           8003
Needs Improvement    7831
Good                 7808
Poor                 5755
Excellent            4854
Extremely Poor       1329
Name: Seat_Comfort, dtype: int64



Value counts for column: Seat_Class
-----------------------------------------
Ordinary     17860
Green Car    17742
Name: Seat_Class, dtype: int64



Value counts for column: Arrival_Time_Convenient
-----------------------------------------
Good                 7361
Excellent            6589
Acceptable           5844
Needs Improvement    5684
Poor                 5131
Extremely Poor       1668
Name: Arrival_Time_Convenient, dtype: int64



Value counts for co

In [26]:
get_value_counts(travel_train_data)

Value counts for column: ID
-----------------------------------------
98800001    1
98862916    1
98862925    1
98862924    1
98862923    1
           ..
98831459    1
98831458    1
98831457    1
98831456    1
98894379    1
Name: ID, Length: 94379, dtype: int64



Value counts for column: Gender
-----------------------------------------
Female    47815
Male      46487
Name: Gender, dtype: int64



Value counts for column: Customer_Type
-----------------------------------------
Loyal Customer       69823
Disloyal Customer    15605
Name: Customer_Type, dtype: int64



Value counts for column: Age
-----------------------------------------
39.0    2688
25.0    2542
40.0    2333
44.0    2291
41.0    2248
        ... 
73.0      47
79.0      39
76.0      38
78.0      30
85.0      18
Name: Age, Length: 75, dtype: int64



Value counts for column: Type_Travel
-----------------------------------------
Business Travel    58617
Personal Travel    26536
Name: Type_Travel, dtype: int64



Value coun

In [27]:
get_value_counts(travel_test_data)

Value counts for column: ID
-----------------------------------------
99900001    1
99923738    1
99923732    1
99923733    1
99923734    1
           ..
99911862    1
99911861    1
99911860    1
99911859    1
99935602    1
Name: ID, Length: 35602, dtype: int64



Value counts for column: Gender
-----------------------------------------
Female    18069
Male      17503
Name: Gender, dtype: int64



Value counts for column: Customer_Type
-----------------------------------------
Loyal Customer       26349
Disloyal Customer     5870
Name: Customer_Type, dtype: int64



Value counts for column: Age
-----------------------------------------
39.0    1003
25.0     968
40.0     877
41.0     842
45.0     820
        ... 
75.0      20
78.0      14
79.0      13
74.0      11
85.0       7
Name: Age, Length: 75, dtype: int64



Value counts for column: Type_Travel
-----------------------------------------
Business Travel    22313
Personal Travel     9841
Name: Type_Travel, dtype: int64



Value coun

## Preprocessing data

In [49]:
def impute_missing_values(df):
    # Create imputers
    numeric_imputer = SimpleImputer(strategy='median')
    categorical_imputer = SimpleImputer(strategy='most_frequent')
    
    # Identify numeric and categorical columns
    numeric_cols = df.select_dtypes(include=['float64', 'int64']).columns
    categorical_cols = df.select_dtypes(include=['object']).columns
    
    # Impute numeric columns
    df[numeric_cols] = numeric_imputer.fit_transform(df[numeric_cols])
    
    # Impute categorical columns
    df[categorical_cols] = categorical_imputer.fit_transform(df[categorical_cols])
    
    return df

In [52]:
survey_train_data = impute_missing_values(survey_train_data)

In [53]:
survey_test_data = impute_missing_values(survey_test_data)

In [54]:
travel_train_data = impute_missing_values(travel_train_data)

In [55]:
travel_test_data = impute_missing_values(travel_test_data)

**Combine training data and testing data**

In [56]:
def merge_dataframes(df1, df2, key_column):
    """
    Merges two DataFrames based on a common key column.
    
    Parameters:
    - df1: First DataFrame
    - df2: Second DataFrame
    - key_column: The column on which to merge the DataFrames
    
    Returns:
    - Merged DataFrame
    """
    
    merged_df = pd.merge(df1, df2, on=key_column, how='inner')
    return merged_df

In [57]:
combined_train_data = merge_dataframes(survey_train_data, travel_train_data, 'ID')

In [58]:
combined_train_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 94379 entries, 0 to 94378
Data columns (total 25 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       94379 non-null  float64
 1   Overall_Experience       94379 non-null  float64
 2   Seat_Comfort             94379 non-null  object 
 3   Seat_Class               94379 non-null  object 
 4   Arrival_Time_Convenient  94379 non-null  object 
 5   Catering                 94379 non-null  object 
 6   Platform_Location        94379 non-null  object 
 7   Onboard_Wifi_Service     94379 non-null  object 
 8   Onboard_Entertainment    94379 non-null  object 
 9   Online_Support           94379 non-null  object 
 10  Ease_of_Online_Booking   94379 non-null  object 
 11  Onboard_Service          94379 non-null  object 
 12  Legroom                  94379 non-null  object 
 13  Baggage_Handling         94379 non-null  object 
 14  CheckIn_Service       

In [59]:
combined_test_data = merge_dataframes(survey_test_data, travel_test_data, 'ID')

In [60]:
combined_test_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 35602 entries, 0 to 35601
Data columns (total 24 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       35602 non-null  float64
 1   Seat_Comfort             35602 non-null  object 
 2   Seat_Class               35602 non-null  object 
 3   Arrival_Time_Convenient  35602 non-null  object 
 4   Catering                 35602 non-null  object 
 5   Platform_Location        35602 non-null  object 
 6   Onboard_Wifi_Service     35602 non-null  object 
 7   Onboard_Entertainment    35602 non-null  object 
 8   Online_Support           35602 non-null  object 
 9   Ease_of_Online_Booking   35602 non-null  object 
 10  Onboard_Service          35602 non-null  object 
 11  Legroom                  35602 non-null  object 
 12  Baggage_Handling         35602 non-null  object 
 13  CheckIn_Service          35602 non-null  object 
 14  Cleanliness           

In [61]:
combined_train_data.to_csv('train_data.csv', index=False)
combined_test_data.to_csv('test_data.csv', index=False)

In [111]:
# Import survey test and train data
train_data = pd.read_csv('train_data.csv')
test_data = pd.read_csv('test_data.csv')

## Feature Engineering

In [112]:
def engineer_features(df):
    # Creating a copy of the dataframe to avoid modifying the original one
    df_engineered = df.copy()
    
    # Average Delay
    df_engineered['Average_Delay'] = (df_engineered['Departure_Delay_in_Mins'] + df_engineered['Arrival_Delay_in_Mins']) / 2.0
    
    # Total Service Quality
    service_columns = ["Seat_Comfort", "Catering", "Onboard_Wifi_Service", "Onboard_Entertainment",
                       "Online_Support", "Ease_of_Online_Booking", "Legroom", "Online_Boarding",
                       "Onboard_Service", "Baggage_Handling", "CheckIn_Service", "Cleanliness"]
    
    # Age Group
    bins_age = [0, 30, 60, float('inf')]
    labels_age = ["Young", "Middle-aged", "Senior"]
    df_engineered['Age_Group'] = pd.cut(df_engineered['Age'], bins=bins_age, labels=labels_age, right=False)
    
    # Distance Group
    bins_distance = [0, 1000, 2000, float('inf')]
    labels_distance = ["Short", "Medium", "Long"]
    df_engineered['Distance_Group'] = pd.cut(df_engineered['Travel_Distance'], bins=bins_distance, labels=labels_distance, right=False)
    
    # Is_Delayed
    df_engineered['Is_Delayed'] = ((df_engineered['Departure_Delay_in_Mins'] > 0) | (df_engineered['Arrival_Delay_in_Mins'] > 0)).astype(int)
    
    # Total_Delay
    df_engineered['Total_Delay'] = df_engineered['Departure_Delay_in_Mins'] + df_engineered['Arrival_Delay_in_Mins']
    
    return df_engineered

In [113]:
engineered_train_data = engineer_features(train_data)

In [114]:
engineered_test_data = engineer_features(test_data)

In [115]:
engineered_train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 94379 entries, 0 to 94378
Data columns (total 30 columns):
 #   Column                   Non-Null Count  Dtype   
---  ------                   --------------  -----   
 0   ID                       94379 non-null  float64 
 1   Overall_Experience       94379 non-null  float64 
 2   Seat_Comfort             94379 non-null  object  
 3   Seat_Class               94379 non-null  object  
 4   Arrival_Time_Convenient  94379 non-null  object  
 5   Catering                 94379 non-null  object  
 6   Platform_Location        94379 non-null  object  
 7   Onboard_Wifi_Service     94379 non-null  object  
 8   Onboard_Entertainment    94379 non-null  object  
 9   Online_Support           94379 non-null  object  
 10  Ease_of_Online_Booking   94379 non-null  object  
 11  Onboard_Service          94379 non-null  object  
 12  Legroom                  94379 non-null  object  
 13  Baggage_Handling         94379 non-null  object  
 14  CheckI

## Encoding 

In [63]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 94379 entries, 0 to 94378
Data columns (total 25 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       94379 non-null  float64
 1   Overall_Experience       94379 non-null  float64
 2   Seat_Comfort             94379 non-null  object 
 3   Seat_Class               94379 non-null  object 
 4   Arrival_Time_Convenient  94379 non-null  object 
 5   Catering                 94379 non-null  object 
 6   Platform_Location        94379 non-null  object 
 7   Onboard_Wifi_Service     94379 non-null  object 
 8   Onboard_Entertainment    94379 non-null  object 
 9   Online_Support           94379 non-null  object 
 10  Ease_of_Online_Booking   94379 non-null  object 
 11  Onboard_Service          94379 non-null  object 
 12  Legroom                  94379 non-null  object 
 13  Baggage_Handling         94379 non-null  object 
 14  CheckIn_Service       

In [64]:
get_value_counts(train_data)

Value counts for column: ID
-----------------------------------------
98800001.0    1
98862916.0    1
98862925.0    1
98862924.0    1
98862923.0    1
             ..
98831459.0    1
98831458.0    1
98831457.0    1
98831456.0    1
98894379.0    1
Name: ID, Length: 94379, dtype: int64



Value counts for column: Overall_Experience
-----------------------------------------
1.0    51593
0.0    42786
Name: Overall_Experience, dtype: int64



Value counts for column: Seat_Comfort
-----------------------------------------
Acceptable           21219
Needs Improvement    20946
Good                 20595
Poor                 15185
Excellent            12971
Extremely Poor        3463
Name: Seat_Comfort, dtype: int64



Value counts for column: Seat_Class
-----------------------------------------
Green Car    47435
Ordinary     46944
Name: Seat_Class, dtype: int64



Value counts for column: Arrival_Time_Convenient
-----------------------------------------
Good                 28504
Excellent    

In [116]:
def encode_data_updated(df):
    # Creating a copy of the dataframe to avoid modifying the original one
    encoded_df = df.copy()
    
    # Ordinal Encoding
    ordinal_columns = [
        "Seat_Comfort", "Arrival_Time_Convenient", "Catering", "Onboard_Wifi_Service", 
        "Onboard_Entertainment", "Ease_of_Online_Booking", "Legroom", "Online_Boarding", 
        "Platform_Location", "Online_Support", "Onboard_Service", "Baggage_Handling", 
        "CheckIn_Service", "Cleanliness"
    ]
    
    # Defining the order for ordinal columns based on common values
    order = [
        ["Extremely Poor", "Poor", "Needs Improvement", "Acceptable", "Good", "Excellent"],
        ['Very Inconvenient', "Inconvenient", "Needs Improvement", "Manageable", "Convenient", "Very Convenient"]
    ]
    
    for col in ordinal_columns:
        if col in ["Platform_Location"]:
            ordinal_encoder = OrdinalEncoder(categories=[order[1]])
        else:
            ordinal_encoder = OrdinalEncoder(categories=[order[0]])
        encoded_df[col] = ordinal_encoder.fit_transform(encoded_df[[col]])
    
    # Label Encoding
    label_columns = ["Seat_Class", "Gender", "Customer_Type", "Type_Travel", "Travel_Class", "Age_Group", "Distance_Group"]
    label_encoder = LabelEncoder()
    for col in label_columns:
        encoded_df[col] = label_encoder.fit_transform(encoded_df[col].astype(str))
    
    return encoded_df

In [117]:
# Test the encoding function
encoded_data = encode_data(engineered_train_data)
encoded_data.head()

ID  Overall_Experience  Seat_Comfort  Seat_Class  \
0  98800001.0                 0.0           2.0           0   
1  98800002.0                 0.0           1.0           1   
2  98800003.0                 1.0           2.0           0   
3  98800004.0                 0.0           3.0           1   
4  98800005.0                 1.0           3.0           1   

   Arrival_Time_Convenient  Catering  Platform_Location  Onboard_Wifi_Service  \
0                      5.0       5.0                5.0                   4.0   
1                      5.0       1.0                2.0                   4.0   
2                      2.0       2.0                2.0                   2.0   
3                      2.0       3.0                2.0                   3.0   
4                      3.0       3.0                3.0                   2.0   

   Onboard_Entertainment  Online_Support  ...  Type_Travel  Travel_Class  \
0                    2.0             3.0  ...            0             0   
1                    1.0             4.0  ...            1             1   
2                    4.0             5.0  ...            0             0   
3                    2.0             3.0  ...            0             0   
4                    4.0             5.0  ...            0             0   

   Travel_Distance  Departure_Delay_in_Mins  Arrival_Delay_in_Mins  \
0            272.0                      0.0                    5.0   
1           2200.0                      9.0                    0.0   
2           1061.0                     77.0                  119.0   
3            780.0                     13.0                   18.0   
4           1981.0                      0.0                    0.0   

   Average_Delay    Age_Group  Distance_Group  Is_Delayed  Total_Delay  
0            2.5  Middle-aged           Short           1          5.0  
1            4.5  Middle-aged            Long           1          9.0  
2           98.0  Middle-aged          Medium           1        196.0  
3           15.5  Middle-aged           Short           1         31.0  
4            0.0  Middle-aged          Medium           0          0.0  

[5 rows x 30 columns]

In [81]:
encoded_data.describe().T.round(2)

count         mean       std         min  \
ID                       94379.0  98847190.00  27245.01  98800001.0   
Overall_Experience       94379.0         0.55      0.50         0.0   
Seat_Comfort             94379.0         2.84      1.39         0.0   
Seat_Class               94379.0         0.50      0.50         0.0   
Arrival_Time_Convenient  94379.0         3.09      1.48         0.0   
Catering                 94379.0         2.87      1.38         0.0   
Platform_Location        94379.0         2.99      1.31         0.0   
Onboard_Wifi_Service     94379.0         3.25      1.32         0.0   
Onboard_Entertainment    94379.0         3.38      1.35         0.0   
Online_Support           94379.0         3.52      1.31         0.0   
Ease_of_Online_Booking   94379.0         3.47      1.31         0.0   
Onboard_Service          94379.0         3.51      1.22         0.0   
Legroom                  94379.0         3.48      1.29         0.0   
Baggage_Handling         94379.0         3.70      1.16         1.0   
CheckIn_Service          94379.0         3.34      1.26         0.0   
Cleanliness              94379.0         3.70      1.15         0.0   
Online_Boarding          94379.0         3.35      1.30         0.0   
Gender                   94379.0         0.49      0.50         0.0   
Customer_Type            94379.0         0.83      0.37         0.0   
Age                      94379.0        39.42     15.11         7.0   
Type_Travel              94379.0         0.28      0.45         0.0   
Travel_Class             94379.0         0.52      0.50         0.0   
Travel_Distance          94379.0      1978.89   1027.96        50.0   
Departure_Delay_in_Mins  94379.0        14.64     38.13         0.0   
Arrival_Delay_in_Mins    94379.0        14.95     38.38         0.0   
Average_Delay            94379.0        14.79     37.86         0.0   
Is_Delayed               94379.0         0.54      0.50         0.0   
Total_Delay              94379.0        29.59     75.72         0.0   

                                25%         50%         75%         max  
ID                       98823595.5  98847190.0  98870784.5  98894379.0  
Overall_Experience              0.0         1.0         1.0         1.0  
Seat_Comfort                    2.0         3.0         4.0         5.0  
Seat_Class                      0.0         0.0         1.0         1.0  
Arrival_Time_Convenient         2.0         3.0         4.0         5.0  
Catering                        2.0         3.0         4.0         5.0  
Platform_Location               2.0         3.0         4.0         5.0  
Onboard_Wifi_Service            2.0         3.0         4.0         5.0  
Onboard_Entertainment           2.0         4.0         4.0         5.0  
Online_Support                  3.0         4.0         5.0         5.0  
Ease_of_Online_Booking          2.0         4.0         5.0         5.0  
Onboard_Service                 3.0         4.0         4.0         5.0  
Legroom                         2.0         4.0         5.0         5.0  
Baggage_Handling                3.0         4.0         5.0         5.0  
CheckIn_Service                 3.0         3.0         4.0         5.0  
Cleanliness                     3.0         4.0         5.0         5.0  
Online_Boarding                 2.0         4.0         4.0         5.0  
Gender                          0.0         0.0         1.0         1.0  
Customer_Type                   1.0         1.0         1.0         1.0  
Age                            27.0        40.0        51.0        85.0  
Type_Travel                     0.0         0.0         1.0         1.0  
Travel_Class                    0.0         1.0         1.0         1.0  
Travel_Distance              1359.0      1923.0      2538.0      6951.0  
Departure_Delay_in_Mins         0.0         0.0        12.0      1592.0  
Arrival_Delay_in_Mins           0.0         0.0        13.0      1584.0  
Average_Delay                   0.0        

In [118]:
test_encoded_data = encode_data(engineered_test_data)

In [121]:
def encode_specific_columns(train_df, test_df):
    # Creating a copy of the dataframes to avoid modifying the original ones
    encoded_train = train_df.copy()
    encoded_test = test_df.copy()

    # Columns to be encoded
    columns_to_encode = ["Age_Group", "Distance_Group"]

    for col in columns_to_encode:
        # Using LabelEncoder to encode the column
        label_encoder = LabelEncoder()
        
        # Fit the encoder using the training data and transform both training and testing data
        encoded_train[col] = label_encoder.fit_transform(encoded_train[col])
        encoded_test[col] = label_encoder.transform(encoded_test[col])

    return encoded_train, encoded_test

In [122]:
encoded_train_data, encoded_test_data = encode_specific_columns(encoded_data, test_encoded_data)

In [123]:
encoded_train_data.to_csv('encoded_train_data.csv', index=False)
encoded_test_data.to_csv('encoded_test_data.csv', index=False)

In [124]:
encoded_train_data = pd.read_csv('encoded_train_data.csv')
encoded_test_data = pd.read_csv('encoded_test_data.csv')

## Normalization

In [153]:
encoded_train_data = pd.read_csv('encoded_train_data.csv')
encoded_test_data = pd.read_csv('encoded_test_data.csv')

In [155]:
# Columns to exclude from normalization
cols_to_exclude_train = ['ID', 'Overall_Experience']
cols_to_exclude_test = ['ID']

# Check for columns in train data but not in test data
missing_cols = set(encoded_train_data.columns) - set(encoded_test_data.columns) - set(cols_to_exclude_train)
for col in missing_cols:
    encoded_test_data[col] = 0  # Add the missing column to the test data and fill with zeros

# Separate out the columns that shouldn't be normalized
train_data_to_exclude = encoded_train_data[cols_to_exclude_train]
test_data_to_exclude = encoded_test_data[cols_to_exclude_test]

# Drop the columns that shouldn't be normalized
encoded_train_data = encoded_train_data.drop(columns=cols_to_exclude_train)
encoded_test_data = encoded_test_data.drop(columns=cols_to_exclude_test)

# Ensure the order of columns in the test data is the same as in train data
encoded_test_data = encoded_test_data[encoded_train_data.columns]

# Initialize the scaler
scaler = StandardScaler()

# Normalize the data
encoded_train_data_normalized = pd.DataFrame(scaler.fit_transform(encoded_train_data), columns=encoded_train_data.columns)
encoded_test_data_normalized = pd.DataFrame(scaler.transform(encoded_test_data), columns=encoded_test_data.columns)

# Add back the columns that shouldn't be normalized
encoded_train_data_normalized = pd.concat([train_data_to_exclude, encoded_train_data_normalized], axis=1)
encoded_test_data_normalized = pd.concat([test_data_to_exclude, encoded_test_data_normalized], axis=1)

# Drop the missing columns from the normalized test data
encoded_test_data_normalized = encoded_test_data_normalized.drop(columns=missing_cols)

# Save the normalized data to new CSV files
encoded_train_data_normalized.to_csv('encoded_train_data_normalized.csv', index=False)
encoded_test_data_normalized.to_csv('encoded_test_data_normalized.csv', index=False)

## Modeling

### Logistic Regression

#### No-Normilization

In [100]:
encoded_train_data = pd.read_csv('encoded_train_data.csv')
encoded_test_data = pd.read_csv('encoded_test_data.csv')

In [125]:
# Drop the 'ID' column and separate the features and target variable for training data
X_train = encoded_train_data.drop(columns=['ID', 'Overall_Experience'])
y_train = encoded_train_data['Overall_Experience']

# Drop the 'ID' column and separate the features and target variable for testing data
X_test = encoded_test_data.drop(columns=['ID'])

In [126]:
# Initialize the Logistic Regression model
logistic_model = LogisticRegression(max_iter=1000)

In [127]:
# Train the model
logistic_model.fit(X_train, y_train)

C:\Users\Grego\AppData\Roaming\Python\Python39\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(max_iter=1000)

In [129]:
# Predict on the training set
y_train_pred = logistic_model.predict(X_train)

# Calculate and display training accuracy
train_accuracy = accuracy_score(y_train, y_train_pred)
print(f"Training Accuracy: {train_accuracy * 100:.2f}%")

# Display a detailed classification report for the training set
print(classification_report(y_train, y_train_pred))

Training Accuracy: 82.12%
              precision    recall  f1-score   support

         0.0       0.81      0.79      0.80     42786
         1.0       0.83      0.85      0.84     51593

    accuracy                           0.82     94379
   macro avg       0.82      0.82      0.82     94379
weighted avg       0.82      0.82      0.82     94379



In [128]:
# Predict on the test set
y_pred = logistic_model.predict(X_test)

#### Normalized

In [156]:
encoded_train_data = pd.read_csv('encoded_train_data_normalized.csv')
encoded_test_data = pd.read_csv('encoded_test_data_normalized.csv')
# Drop the 'ID' column and separate the features and target variable for training data
X_train = encoded_train_data.drop(columns=['ID', 'Overall_Experience'])
y_train = encoded_train_data['Overall_Experience']

# Drop the 'ID' column and separate the features and target variable for testing data
X_test = encoded_test_data.drop(columns=['ID'])
# Initialize the Logistic Regression model
logistic_model = LogisticRegression(max_iter=1000)
# Train the model
logistic_model.fit(X_train, y_train)
# Predict on the training set
y_train_pred = logistic_model.predict(X_train)

# Calculate and display training accuracy
train_accuracy = accuracy_score(y_train, y_train_pred)
print(f"Training Accuracy: {train_accuracy * 100:.2f}%")

# Display a detailed classification report for the training set
print(classification_report(y_train, y_train_pred))
# Predict on the test set
y_pred = logistic_model.predict(X_test)

Training Accuracy: 83.21%
              precision    recall  f1-score   support

         0.0       0.82      0.81      0.81     42786
         1.0       0.84      0.85      0.85     51593

    accuracy                           0.83     94379
   macro avg       0.83      0.83      0.83     94379
weighted avg       0.83      0.83      0.83     94379



### Random Forest

In [182]:
encoded_train_data = pd.read_csv('encoded_train_data.csv')
encoded_test_data = pd.read_csv('encoded_test_data.csv')
# Drop the 'ID' column and separate the features and target variable for training data
X_train = encoded_train_data.drop(columns=['ID', 'Overall_Experience'])
y_train = encoded_train_data['Overall_Experience']

# Drop the 'ID' column and separate the features and target variable for testing data
X_test = encoded_test_data.drop(columns=['ID'])

In [183]:
from sklearn.ensemble import RandomForestClassifier

In [184]:
# Train a Random Forest classifier
rf_model = RandomForestClassifier()
rf_model.fit(X_train, y_train)

RandomForestClassifier()

In [185]:
# Return the model's training accuracy
train_accuracy = rf_model.score(X_train, y_train)
train_accuracy

0.999978808845188

In [186]:
# Predict on the training set
y_train_pred = rf_model.predict(X_train)
# Calculate and display training accuracy
train_accuracy = accuracy_score(y_train, y_train_pred)
print(f"Training Accuracy: {train_accuracy * 100:.2f}%")

# Display a detailed classification report for the training set
print(classification_report(y_train, y_train_pred))

Training Accuracy: 100.00%
              precision    recall  f1-score   support

         0.0       1.00      1.00      1.00     42786
         1.0       1.00      1.00      1.00     51593

    accuracy                           1.00     94379
   macro avg       1.00      1.00      1.00     94379
weighted avg       1.00      1.00      1.00     94379



In [187]:
y_pred = rf_model.predict(X_test)
# Convert the 'float_column' to an integer
encoded_test_data['ID'] = encoded_test_data['ID'].astype(int)

# Create a new dataframe with 'ID' and 'Overall_Experience'
predicted_df = pd.DataFrame({
    'ID': encoded_test_data['ID'],
    'Overall_Experience': y_pred
})

predicted_df.head()

ID  Overall_Experience
0  99900001                 1.0
1  99900002                 1.0
2  99900003                 1.0
3  99900004                 0.0
4  99900005                 1.0

In [188]:
predicted_df.to_csv('Submission2.csv', index=False)

### XGBOOST

In [20]:
import xgboost as xgb


#### Normalization

In [167]:
encoded_train_data = pd.read_csv('encoded_train_data_normalized.csv')
encoded_test_data = pd.read_csv('encoded_test_data_normalized.csv')
# Drop the 'ID' column and separate the features and target variable for training data
X_train = encoded_train_data.drop(columns=['ID', 'Overall_Experience'])
y_train = encoded_train_data['Overall_Experience']

# Drop the 'ID' column and separate the features and target variable for testing data
X_test = encoded_test_data.drop(columns=['ID'])

In [168]:
# Train the XGBoost classifier with adjusted parameters and evaluate its performance on the training set
xgb_model = xgb.XGBClassifier(max_depth=6, n_estimators=100, subsample=0.8, use_label_encoder=False)
xgb_model.fit(X_train, y_train)

# Calculate the training accuracy
xgb_train_accuracy = xgb_model.score(X_train, y_train)
xgb_train_accuracy

0.9696330751544305

In [169]:
# Predict on the training set
y_train_pred = xgb_model.predict(X_train)
# Calculate and display training accuracy
train_accuracy = accuracy_score(y_train, y_train_pred)
print(f"Training Accuracy: {train_accuracy * 100:.2f}%")

# Display a detailed classification report for the training set
print(classification_report(y_train, y_train_pred))

Training Accuracy: 96.96%
              precision    recall  f1-score   support

         0.0       0.96      0.97      0.97     42786
         1.0       0.98      0.97      0.97     51593

    accuracy                           0.97     94379
   macro avg       0.97      0.97      0.97     94379
weighted avg       0.97      0.97      0.97     94379



In [ ]:
# Create a new dataframe with 'ID' and 'Overall_Experience'
predicted_df = pd.DataFrame({
    'ID': encoded_test_data['ID'],
    'Overall_Experience': y_pred
})

predicted_df.head()

ID  Overall_Experience
0  99900001.0                   1
1  99900002.0                   1
2  99900003.0                   1
3  99900004.0                   0
4  99900005.0                   1

In [ ]:
predicted_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35602 entries, 0 to 35601
Data columns (total 2 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  35602 non-null  float64
 1   Overall_Experience  35602 non-null  int32  
dtypes: float64(1), int32(1)
memory usage: 417.3 KB


In [ ]:


predicted_df.to_csv('xgb.csv', index=False)

#### No-Normalization

In [166]:
encoded_train_data = pd.read_csv('encoded_train_data.csv')
encoded_test_data = pd.read_csv('encoded_test_data.csv')
# Drop the 'ID' column and separate the features and target variable for training data
X_train = encoded_train_data.drop(columns=['ID', 'Overall_Experience'])
y_train = encoded_train_data['Overall_Experience']

# Drop the 'ID' column and separate the features and target variable for testing data
X_test = encoded_test_data.drop(columns=['ID'])

In [163]:
# Train the XGBoost classifier with adjusted parameters and evaluate its performance on the training set
xgb_model = xgb.XGBClassifier(max_depth=6, n_estimators=100, subsample=0.8, use_label_encoder=False)
xgb_model.fit(X_train, y_train)

# Calculate the training accuracy
xgb_train_accuracy = xgb_model.score(X_train, y_train)
xgb_train_accuracy

0.9696330751544305

In [165]:
# Predict on the training set
y_train_pred = xgb_model.predict(X_train)
# Calculate and display training accuracy
train_accuracy = accuracy_score(y_train, y_train_pred)
print(f"Training Accuracy: {train_accuracy * 100:.2f}%")

# Display a detailed classification report for the training set
print(classification_report(y_train, y_train_pred))

Training Accuracy: 96.96%
              precision    recall  f1-score   support

         0.0       0.96      0.97      0.97     42786
         1.0       0.98      0.97      0.97     51593

    accuracy                           0.97     94379
   macro avg       0.97      0.97      0.97     94379
weighted avg       0.97      0.97      0.97     94379



In [170]:
y_pred = xgb_model.predict(X_test)

In [178]:
# Convert the 'float_column' to an integer
encoded_test_data['ID'] = encoded_test_data['ID'].astype(int)

# Create a new dataframe with 'ID' and 'Overall_Experience'
predicted_df = pd.DataFrame({
    'ID': encoded_test_data['ID'],
    'Overall_Experience': y_pred
})

predicted_df.head()

ID  Overall_Experience
0  99900001                   1
1  99900002                   1
2  99900003                   1
3  99900004                   0
4  99900005                   1

In [180]:
predicted_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35602 entries, 0 to 35601
Data columns (total 2 columns):
 #   Column              Non-Null Count  Dtype
---  ------              --------------  -----
 0   ID                  35602 non-null  int32
 1   Overall_Experience  35602 non-null  int32
dtypes: int32(2)
memory usage: 278.3 KB


In [181]:
predicted_df.to_csv('Submission1.csv', index=False)

## Model Pipeline Script 

In [15]:
from sklearn.metrics import accuracy_score, classification_report
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
import pandas as pd
from sklearn.model_selection import RandomizedSearchCV
import numpy as np
import joblib



In [16]:
encoded_train_data = pd.read_csv('encoded_train_data_normalized.csv')
encoded_test_data = pd.read_csv('encoded_test_data_normalized.csv')
# Drop the 'ID' column and separate the features and target variable for training data
X_train = encoded_train_data.drop(columns=['ID', 'Overall_Experience'])
y_train = encoded_train_data['Overall_Experience']

# Drop the 'ID' column and separate the features and target variable for testing data
X_test = encoded_test_data.drop(columns=['ID'])

In [17]:
def randomized_search(classifiers, param_distributions, X_train, y_train):
    best_classifiers = {}
    for name, clf in classifiers:
        print(f'------CLASSIFIER: {name} STARTING-------')
        search = RandomizedSearchCV(clf, param_distributions=param_distributions[name],
                                    n_iter=50, scoring='accuracy', n_jobs=-1, cv=5, random_state=0)
        print('Fitting')
        search.fit(X_train, y_train)
        print('Fit')
        
        # Save the best model for this classifier
        best_model_path = f"{name.replace(' ', '_')}_best_model.pkl"
        joblib.dump(search.best_estimator_, best_model_path)
        print(f'Saved best {name} model to {best_model_path}')
        
        best_classifiers[name] = search.best_estimator_
        print(f'------CLASSIFIER: {name} ENDING-------')
    return best_classifiers

def get_best_classifier_from_search(best_classifiers, X_train, y_train):
    print('Choosing best classifier')
    scores = {}
    for name, clf in best_classifiers.items():
        scores[name] = accuracy_score(y_train, clf.predict(X_train))
    best_classifier_name = max(scores, key=scores.get)
    best_classifier = best_classifiers[best_classifier_name]
    
    # Save the overall best classifier
    best_model_path = f"Overall_best_{best_classifier_name.replace(' ', '_')}.pkl"
    joblib.dump(best_classifier, best_model_path)
    print(f'Saved overall best model to {best_model_path}')
    
    return best_classifier_name, best_classifier

def predict_and_save_best_model(best_classifier, X_test, encoded_test_data):
    print('Saving Data')
    # Predict on the X_test data
    y_pred = best_classifier.predict(X_test)
    
    # Convert the 'float_column' to an integer
    encoded_test_data['ID'] = encoded_test_data['ID'].astype(int)

    # Create a new dataframe with 'ID' and 'Overall_Experience'
    predicted_df = pd.DataFrame({
        'ID': encoded_test_data['ID'],
        'Overall_Experience': y_pred
    })
    predicted_df.to_csv('Submission3.csv', index=False)


In [18]:
param_distributions = {
    "XGBoost": {
        'learning_rate': [0.01, 0.1],
        'n_estimators': [50, 100],
        'max_depth': [3, 5, 7],
        'subsample': [0.8, 1.0],
        'colsample_bytree': [0.8, 1.0],
        'gamma': [0, 0.1, 0.2]
    }
}

In [21]:
classifiers = [("XGBoost", xgb.XGBClassifier(random_state=0, use_label_encoder=False))]


In [22]:
# Perform Randomized Search
best_classifiers = randomized_search(classifiers, param_distributions, X_train, y_train)

# Get the best classifier from the search results
best_classifier_name, best_classifier = get_best_classifier_from_search(best_classifiers, X_train, y_train)

# Predict and save results using the best classifier
predict_and_save_best_model(best_classifier, X_test, encoded_test_data)

------CLASSIFIER: XGBoost STARTING-------
Fitting
Fit
Saved best XGBoost model to XGBoost_best_model.pkl
------CLASSIFIER: XGBoost ENDING-------
Choosing best classifier
Saved overall best model to Overall_best_XGBoost.pkl
Saving Data


In [8]:
def load_model_and_predict(filename, X_test, encoded_test_data):
    """
    Load a saved model from a specified file and use it to make predictions on the provided data.
    
    Parameters:
    - filename: The name of the file from which to load the model.
    - X_test: The test data to make predictions on.
    - encoded_test_data: The encoded test data to construct the final prediction dataframe.
    
    Returns:
    - predicted_df: A dataframe with the predictions.
    """
    # Load the saved model
    loaded_model = joblib.load(filename)
    
    # Predict using the loaded model
    y_pred = loaded_model.predict(X_test)
    
    # Convert the 'float_column' to an integer
    encoded_test_data['ID'] = encoded_test_data['ID'].astype(int)

    # Create a dataframe with predictions
    predicted_df = pd.DataFrame({
        'ID': encoded_test_data['ID'],
        'Overall_Experience': y_pred
    })
    
    # Optionally, save to a CSV file
    predicted_df.to_csv('LoadedModel_Submission.csv', index=False)
    
    print(f"Predictions saved to LoadedModel_Submission.csv")
    
    return predicted_df

In [9]:
load_model_and_predict('Decision_Tree_best_model.pkl', X_test, encoded_test_data)

Predictions saved to LoadedModel_Submission.csv


ID  Overall_Experience
0      99900001                 1.0
1      99900002                 1.0
2      99900003                 1.0
3      99900004                 0.0
4      99900005                 1.0
...         ...                 ...
35597  99935598                 0.0
35598  99935599                 1.0
35599  99935600                 0.0
35600  99935601                 1.0
35601  99935602                 0.0

[35602 rows x 2 columns]

In [14]:
load_model_and_predict('K-Nearest_Neighbors_best_model.pkl', X_test, encoded_test_data)

AttributeError: 'Flags' object has no attribute 'c_contiguous'

In [12]:
load_model_and_predict('Random_Forest_best_model.pkl', X_test, encoded_test_data)

Predictions saved to LoadedModel_Submission.csv


ID  Overall_Experience
0      99900001                 1.0
1      99900002                 1.0
2      99900003                 1.0
3      99900004                 0.0
4      99900005                 1.0
...         ...                 ...
35597  99935598                 0.0
35598  99935599                 1.0
35599  99935600                 0.0
35600  99935601                 1.0
35601  99935602                 0.0

[35602 rows x 2 columns]

In [23]:
load_model_and_predict('Overall_best_XGBoost.pkl', X_test, encoded_test_data)

Predictions saved to LoadedModel_Submission.csv


ID  Overall_Experience
0      99900001                   1
1      99900002                   1
2      99900003                   1
3      99900004                   0
4      99900005                   1
...         ...                 ...
35597  99935598                   0
35598  99935599                   1
35599  99935600                   1
35600  99935601                   1
35601  99935602                   0

[35602 rows x 2 columns]

## Submission

In [175]:
sample = pd.read_csv('Sample_Submission_.csv')

In [176]:
sample.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35602 entries, 0 to 35601
Data columns (total 2 columns):
 #   Column              Non-Null Count  Dtype
---  ------              --------------  -----
 0   ID                  35602 non-null  int64
 1   Overall_Experience  35602 non-null  int64
dtypes: int64(2)
memory usage: 556.4 KB
